src/proxy/connection_manager.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::proxy::error;
+use crate::rbac::Connection;
+use drain;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::RwLock;
+
+pub struct ConnectionDrain {
+    tx: drain::Signal,
+    rx: drain::Watch,
+}
+
+impl ConnectionDrain {
+    fn new() -> Self {
+        let (tx, rx) = drain::channel();
+        ConnectionDrain { tx, rx }
+    }
+
+    /// drain drops the internal reference to rx and then signals drain on the tx
+    // always inline, this is for convenience so that we don't forget do drop the rx but there's really no reason it needs to grow the stack
+    #[inline(always)]
+    async fn drain(self) {
+        drop(self.rx); // very important, drain cannont complete if there are outstand rx
+        self.tx.drain().await;
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionManager {
+    drains: Arc<RwLock<HashMap<Connection, ConnectionDrain>>>,
+}
+
+impl ConnectionManager {
+    pub fn new() -> Self {
+        ConnectionManager {
+            drains: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    pub async fn track(self, c: &Connection) -> drain::Watch {
+        // consider removing this whole if let since if it's None we need to perform another get inside the write lock to prevnt racy inserts
+        if let Some(cd) = self.drains.read().await.get(c) {
+            return cd.rx.clone();
+        }
+        let cd = ConnectionDrain::new();
+        let rx = cd.rx.clone();
+        //TODO: this was racy, another insert may happen between dropping the read lock and attaining this write lock
+        // try_insert is the best solution once it's no longer a nightly-only experimental API
+        let mut drains = self.drains.write().await;
+        if let Some(w) = drains.get(c) {
+            return w.rx.clone();
+        }
+        drains.insert(c.clone(), cd);
+        rx
+    }
+
+    pub async fn drain(&self, c: &Connection) {
```

Comment:
is there a graceful draining time?

Reply:
That's still TBD

Reply:
I think there are really 2 behaviors we want here:
1. drain: used when ztunnel is shutting down. sends goaway and continues to handle traffic
2. close: used when auth policy changes and existing connections become denied. stop handling traffic

Reply:
@howardjohn and @kyessenov, does this approach seem reasonable?

Reply:
I don't see any drain here. This is just local close?

Reply:
As it sits now even when policy changes from allow to deny we send a goaway and continue to handle traffic. We'd originally implemented drains in this way because our only connection drains were when ztunnel was being shutdown and in that case we decided that k8s killing the pod would be our drain timeout.

From an auth deny perspective this didn't seem correct though. I'm just about ready to push some new code which has both drain (for shutdown) and close (for auth) implemented for inbound.

Reply:
Let's not conflate HBONE goaway with application drain. HBONE goaway is an infastructure mechanism and should not be used for application intent - it's like reinstalling a cluster when redeploying an app.

For application policies, there's no drain or GOAWAY here AFAICT. You should just close the connection.

Reply:
By close, I mean RST. I think we shouldn't do half close here.